presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ImplementExceptAsUnion.java

```diff
@@ -0,0 +1,102 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.ExpressionUtils;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.ExceptNode;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.UnionNode;
+import io.prestosql.sql.tree.ComparisonExpression;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.GenericLiteral;
+
+import java.util.List;
+
+import static com.google.common.collect.Iterables.concat;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.sql.planner.iterative.rule.SetOperationNodeUtil.allocateSymbols;
+import static io.prestosql.sql.planner.iterative.rule.SetOperationNodeUtil.appendMarkers;
+import static io.prestosql.sql.planner.iterative.rule.SetOperationNodeUtil.computeCounts;
+import static io.prestosql.sql.planner.iterative.rule.SetOperationNodeUtil.project;
+import static io.prestosql.sql.planner.iterative.rule.SetOperationNodeUtil.union;
+import static io.prestosql.sql.planner.plan.Patterns.except;
+import static io.prestosql.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static io.prestosql.sql.tree.ComparisonExpression.Operator.GREATER_THAN_OR_EQUAL;
+
+/**
+ * Converts INTERSECT  queries into UNION ALL..GROUP BY...WHERE
+ * Eg:  SELECT a FROM foo EXCEPT SELECT x FROM bar
+ * <p/>
+ * =>
+ * <p/>
+ * SELECT a
+ * FROM
+ * (SELECT a,
+ * COUNT(foo_marker) AS foo_cnt,
+ * COUNT(bar_marker) AS bar_cnt
+ * FROM
+ * (
+ * SELECT a, true as foo_marker, null as bar_marker FROM foo
+ * UNION ALL
+ * SELECT x, null as foo_marker, true as bar_marker FROM bar
+ * ) T1
+ * GROUP BY a) T2
+ * WHERE foo_cnt >= 1 AND bar_cnt = 0;
+ */
+public class ImplementExceptAsUnion
+        implements Rule<ExceptNode>
+{
+    private static final String MARKER = "marker";
+
+    private static final Pattern<ExceptNode> PATTERN = except();
+
+    @Override
+    public Pattern<ExceptNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(ExceptNode node, Captures captures, Context context)
+    {
+        List<Symbol> markers = allocateSymbols(context.getSymbolAllocator(), node.getSources().size(), MARKER, BOOLEAN);
```

Comment:
The only difference between how Intersect and Except are implemented is in how the final WHERE clause is constructed (i.e., `foo_cnt >= 1 AND bar_cnt = 0` in the example above). I'd structure it in a different way:

- Move the logic for building the subplan for `T2` into the common utility class. It should return a composite object with the plan and the symbols representing the counts for each of the branches of the set operation. 
- Add the operation-specific logic on top of it inside the rule